PROGRAMS.md

```diff
@@ -0,0 +1,223 @@
+3. The program itself must accept the attachment through the `onAttachProgram` hook
+
+This multi-layered validation ensures that only appropriate programs can be attached to entities, maintaining the integrity of the world.
+
+### 3.4 Gas Safety Mechanisms
+
+To prevent malicious or poorly implemented programs from consuming excessive gas, calls with a fixed amount of gas are used. Safe calls are only used for hooks that correspond to actions that the entity might be incentivized to prevent (e.g. `onHit`).
+
+The `SAFE_PROGRAM_GAS` constant (set to 1,000,000 gas units) limits the gas available to safe program calls, preventing the program from consuming all the gas available in the transaction.
+
+
+## 4. Program Execution in Context
+
+### 4.1 Machine Interactions
+
+Programs can respond to interactions with machines, such as when a player fuels a machine:
+
+```solidity
+function fuelMachine(EntityId callerEntityId, EntityId machineEntityId, uint16 fuelAmount) public {
+  // Validation and fuel logic
+  // ...
+
+  // Call program hook
+  ProgramId program = baseEntityId.getProgram();
+  program.callOrRevert(abi.encodeCall(IFuelHook.onFuel, (callerEntityId, baseEntityId, fuelAmount, "")));
+
+  // Notification logic
+  // ...
+}
+```
+
+This allows machines to execute custom logic when they receive fuel, potentially triggering complex behaviors or state changes.
+
+## 5. Practical Applications
+
+### 5.1 Smart Objects
+
+Programs enable the creation of smart objects with custom behaviors:
+
+- **Smart Doors**: Doors that only open for specific entities or under certain conditions
+- **Resource Generators**: Machines that produce resources at regular intervals
+- **Security Systems**: Force fields that restrict actions based on complex rules
+- **Interactive NPCs**: Non-player characters with programmable behavior
+
+### 5.2 Gameplay Mechanics
+
+Programs can implement complex gameplay mechanics:
+
+- **Quest Systems**: Programs that track player progress and reward completion
+- **Economic Systems**: Trading posts with dynamic pricing algorithms
+- **Combat Systems**: Custom combat mechanics for specific areas or entities
+- **Crafting Systems**: Advanced crafting pipelines that reward resource providers and periodically craft specific items
+
+### 5.3 World Governance
+
+Programs can facilitate governance mechanisms:
+
+- **Access Control**: Programs that manage permissions for different areas
+- **Resource Management**: Systems that regulate resource distribution
+- **Voting Systems**: Mechanisms for collective decision-making
+- **Taxation Systems**: Automated collection and distribution of resources
+
+## 6. Implementation Guide
+
+### 6.1 Creating a Program (WIP)
+
+Example of a simple door program:
+
+```solidity
+contract SimpleDoorProgram is Program {
+  constructor(IWorld world) Program(world) {}
+
+  function onAttach(EntityId caller, EntityId target, bytes memory extraData) external onlyWorld {
```

Comment:
nit: was there a decision around renaming `extraData` to something more descriptive? (ie `programContext` or sth?)
```suggestion
  function onAttach(EntityId caller, EntityId target, bytes memory programContext) external onlyWorld {
```

Reply:
no decision but yeah I think we could do it, will leave for another pr so we can merge this soon